cmd/prometheus/main.go

```diff
@@ -288,6 +292,17 @@ func main() {
 
 	var g group.Group
 	{
+		g.Add(
+			func() error {
+				return discoveryManager.Run()
+			},
+			func(err error) {
+				level.Info(logger).Log("msg", "Stopping target manager...")
+				cancelCtx()
+				level.Info(logger).Log("msg", "Target manager stopped")
```

Comment:
I think those "stopped" logs should be coming after the run function terminated.
Like this, both log lines will be emitted right after each other. That does not tell us whether the component might actually be hanging on shutdown.